### PR TITLE
add a new filter goaway which could send GOAWAY probabilistically to help balance HTTP2 requests

### PR DESCRIPTION
**Pull request by answer1991**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

API Server and client use HTTP2 to connect each other, which means client always hit the same API Server across different request and the requests are NOT balanced even if we use a load-balance to expose API Server. 

With this PR merged, a new filter `goaway` could send GOAWAY to HTTP2 client probabilistically. After client received GOAWAY, the in-flight watch request will NOT be influenced, and the new requests will use a new TCP connection, which means load-balance will work after client receive GOAWAY

**Which issue(s) this PR fixes**:

None

**Special notes for your reviewer**:

Nothing yet.

**Does this PR introduce a user-facing change?**:

```release-note
Apiserver add a new flag --goaway-chance which is the fraction of requests that will be closed gracefully(GOAWAY) to prevent HTTP/2 clients from getting stuck on a single apiserver. 
After the connection closed(received GOAWAY), the client's other in-flight requests won't be affected, and the client will reconnect. 
The flag min value is 0 (off), max is .02 (1/50 requests); .001 (1/1000) is a recommended starting point.
Clusters with single apiservers, or which don't use a load balancer, should NOT enable this.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

None

### CODE REVIEW

**Comment by answer1991** [comment]
/hold

let me do some more test firstly

**Comment by answer1991** [comment]
/hold cancel

The testing result is awesome. After client received GOAWAY, the in-flight watch request will NOT be influenced, and the new requests will use a new TCP connection, which means load-balance will work after client receive GOAWAY. Please see details and result in my test cases.

**Comment by answer1991** [comment]
cc @lavalamp @wojtek-t @deads2k

**Review by dixudx — APPROVED** [review]
/lgtm

File: staging/src/k8s.io/apiserver/pkg/server/filters/maxinflight.go
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/filters/maxinflight.go]: cc @caesarxuchao @smarterclayton

**Comment by answer1991** [comment]
BTW, I'd like to add same action to timeout filter when 504 occur.

**Comment by smarterclayton** [comment]
This is great to see - https://github.com/kubernetes/kubernetes/issues/48610 and its fix https://github.com/kubernetes/kubernetes/pull/48616 were blocked on Golang support before, so I'll help review and drive this to conclusion.

File: staging/src/k8s.io/apiserver/pkg/server/filters/maxinflight.go
Comment (smarterclayton) [reply staging/src/k8s.io/apiserver/pkg/server/filters/maxinflight.go]: This needs to be probabilistic - tearing down every connection would be too expensive.  See the linked #48616 - I believe this needs to be much closer to that.

Comment (lavalamp) [reply staging/src/k8s.io/apiserver/pkg/server/filters/maxinflight.go]: I'd go further-- once the last apiserver becomes overloaded you end up with a complete system failure as every client request becomes drastically more expensive.

It's much, much better for us to queue requests. 1.18 has a useful alpha: https://github.com/kubernetes/kubernetes/pull/88177#issuecomment-588945806

I think best is for clients to occasionally do this, the client is actually in a better position to figure out if additional attempts have a good chance of finding a not-busy apiserver.

If we do this on the server, it should have a low probability, like 1/100.

We should never ever do this in a single-apiserver system.

**Comment by lavalamp** [comment]
/hold

The current state of this would increase load on overloaded clusters, making it hard for them to recover.

**Comment by lavalamp** [comment]
Personally, I'd rather get more eyes on the priority and fairness effort than spend a lot of time fixing the current max-in-flight limit. Currently we are working on observability, testing, and documentation.

I'm not sure how to make this GOAWAY action safe for overloaded clusters (without apiservers communicating their load level), given that fresh connections are much heavier than reused ones.

**Comment by smarterclayton** [comment]
This needs the probabilistic load shedding, not the static load shedding.

I disagree with the hold here - connection stickiness is orthogonal to traffic shaping.

**Comment by lavalamp** [comment]
I can live with this if the probability of a drop is inversely proportional to the server's over-loaded-ness.

That would make the most loaded apiservers do most of the drops.

**Comment by deads2k** [comment]
I think I'd rather see a reasonable rebalance regardless of the tooManyRequests or not.

Given three kube-apiservers that restart one at a time, we see behavior like this:
1. master1 drops all load during a graceful shutdown
2. clients re-establish connections to master2 or master3
3. master1 starts back up with almost no load
4. master2 drops all load during a graceful shutdown
5. repeat

the end result is that master3 carries comparatively low load and it never truly settles back out

If we disconnected any connection on some low chance (1/1000?), then clients would have a safe signal to try again (their request did nothing). This means that client-go could auto-retry  for instance, giving a seamless experience for callers.  It would also allow rebalancing over time by ensuring no individual connection stays alive for interminable timeframes.

**Comment by lavalamp** [comment]
I can also agree with a 1/50th (or rarer) fixed chance of a drop. (but never in a single apiserver system)

I don't think we have a test that runs a cluster sufficiently close to the edge that we'd know if this breaks it or not.

**Comment by lavalamp** [comment]
I like the idea of just always dropping with a fixed, very low percentage. We can combine that with a slightly higher percentage when loaded.

If it is low enough it may not be worth it to exclude single apiserver clusters.

**Comment by smarterclayton** [comment]
Yeah, always drop with low percentage of request rate is reasonable.  Reusing connections is just amortizing connection setup cost, and with a low probability that's fine.

**Comment by answer1991** [comment]
Thanks a lot for the comments. 

Let me cherry-pick #48616 and make it probabilistic

**Comment by smarterclayton** [comment]
#48616 is still 429 limited, so for the overall probalistic you'd probably end up moving it higher up the stack (maybe into timeout)

**Comment by lavalamp** [comment]
Add a new handler (very early in the stack) for constant low drop rate.

**Comment by answer1991** [comment]
I think I have finished the work, not sure anything I missing or not cared.

Thanks a lot for review.

File: staging/src/k8s.io/apiserver/pkg/server/config.go
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/server/config.go]: Max value? I suggest .02.

File: staging/src/k8s.io/apiserver/pkg/server/filters/goaway.go
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/server/filters/goaway.go]: This takes a global lock, which is probably not OK.

File: staging/src/k8s.io/apiserver/pkg/server/filters/goaway_test.go
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/server/filters/goaway_test.go]: this could return negative numbers, which rand.Float64() cannot. It's probably OK but just mentioning.

File: staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go]: It really needs to be quite small to not break the cluster. Personally I don't think I'd let it be larger than 1/500; I'm OK with validation permitting values that aren't a good idea, but I can't imagine running with this over 1/50 would be a good idea. Certainly values approaching 1 are really not good and so 1 is definitely too high..

File: staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go]: Mention the max.

File: staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go]: Rewrite this:
`To prevent HTTP/2 clients from getting stuck on a single apiserver, randomly close a connection (GOAWAY). The client's other in-flight requests won't be affected, and the client will reconnect, likely landing on a different apiserver after going through the load balancer again. This argument sets the fraction of requests that will be sent a GOAWAY. Clusters with single apiservers, or which don't use a load balancer, should NOT enable this. Min is 0 (off), Max is .02 (1/50 requests); .001 (1/1000) is a recommended starting point.`

Comment (lavalamp) [reply staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go]: @deads2k for fact checking.

File: staging/src/k8s.io/apiserver/pkg/server/filters/goaway.go
Comment (answer1991) [reply staging/src/k8s.io/apiserver/pkg/server/filters/goaway.go]: Done, construct a new random source

File: staging/src/k8s.io/apiserver/pkg/server/filters/goaway_test.go
Comment (answer1991) [reply staging/src/k8s.io/apiserver/pkg/server/filters/goaway_test.go]: done

File: staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go
Comment (answer1991) [reply staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go]: has add validation to limit max value less than 0.02

File: staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go
Comment (answer1991) [reply staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go]: done

File: staging/src/k8s.io/apiserver/pkg/server/config.go
Comment (answer1991) [reply staging/src/k8s.io/apiserver/pkg/server/config.go]: done

**Comment by answer1991** [comment]
I added a new test case: 

API Server response watch request with GOAWAY will not affect client watching body read.

File: staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go
Comment (deads2k) [reply staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go]: > @deads2k for fact checking.

Looks accurate to me.  I think we can reserve the right to change the default based on experience, but 1/1000 seems like a good starting point for the clusters that will care about this.

File: staging/src/k8s.io/apiserver/pkg/server/filters/goaway.go
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/server/filters/goaway.go]: I think this is now unlocked and can't be used from multiple threads. (and if it is locked, it's still a global chokepoint)

I think you need to make a sync.Pool of these; it attempts to solve this problem and is easy. (see: https://golang.org/src/sync/pool.go?s=3925:3957#L114)

Comment (answer1991) [reply staging/src/k8s.io/apiserver/pkg/server/filters/goaway.go]: yes, `rand.Rand` used in multi-goroutines  without locker would panic. I ran a problematic test before with a wrapper `fmt.Println`: `fmt.Println(rnd.Float64())`, which misleading me.

I had already fixed it using a pool to get a rand.Rand, many thanks for your guide.

File: staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go
Comment (answer1991) [reply staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go]: I agreed 1/1000 is a good start point too. From the perspective of the client, it set up a new connection after 1000 times visit API Server.  It not only takes care of the performance of client reuse connection, but also ensures that the load of API Server(s) is balanced timely.

File: staging/src/k8s.io/apiserver/pkg/server/filters/goaway.go
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/server/filters/goaway.go]: You have to use a random seed, otherwise rejections will be correlated.

Comment (answer1991) [reply staging/src/k8s.io/apiserver/pkg/server/filters/goaway.go]: I think it's not a big problem, as different P/goroutine get a same seed rand but invoke this function does not at the same time. 

Anyway I has changed it to be a random seed which may be more perfect, thanks a lot.

File: staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go]: nit: s/and/or/

File: staging/src/k8s.io/apiserver/pkg/server/filters/goaway.go
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/server/filters/goaway.go]: Personally I'd probably have used a defer (not asking for a change, just noting).

File: staging/src/k8s.io/apiserver/pkg/server/filters/goaway.go
Comment (lavalamp) [reply staging/src/k8s.io/apiserver/pkg/server/filters/goaway.go]: Thanks for the fix; correlated random number streams are just a subtle source of bug that it's best to avoid.

E.g. if you imagine that all the apiservers end up with the same number of these in the pool, create at least some of them at the same time, and get about the same number of requests, then you'll end up with bursts where all the apiservers reject a request very near in time.

**Comment by lavalamp** [comment]
Can we do an e2e run with this turned on before we merge? I just want to make sure that clients all reconnect as expected.

**Comment by lavalamp** [comment]
Ah, I see the request is still handled, the client will just close it afterwards. So, it should be fine.

/lgtm
/approve

**Comment by lavalamp** [comment]
Can you edit the release note to speak about the flag rather than the filter? Thanks.

**Comment by answer1991** [comment]
updated release note.

/hold cancel

**Comment by answer1991** [comment]
Merge progress seems blocked, let me fix the validating error message nit and push again to trigger the merge progress

File: staging/src/k8s.io/apiserver/pkg/server/filters/goaway.go
Comment (answer1991) [reply staging/src/k8s.io/apiserver/pkg/server/filters/goaway.go]: yes, agreed

File: staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go
Comment (answer1991) [reply staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go]: fixed

**Comment by answer1991** [comment]
@lavalamp  kindly ping for a `LGTM` label. I removed the `/hold` label seems to trigger an authorize issue.

BTW, fix the validating error message nit.

> Can we do an e2e run with this turned on before we merge

Not sure how to config the e2e API Server flags.

**Comment by pingstar** [comment]
Can I set it up for different resources?

**Comment by answer1991** [comment]
> Can I set it up for different resources?

It's HTTP handler a filter and not care about resources.

**Pull request merged**
